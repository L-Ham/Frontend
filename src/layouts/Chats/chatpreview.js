import React from 'react';
import PropTypes from 'prop-types';
/*eslint-disable*/
function ChatPreview({sender, lastMessage, senderImageURL, date, onSelect, unreadTotal}) {
    const handleClick = (e) => {
        e.preventDefault();
        e.stopPropagation();
        onSelect();
        // console.log('Sender: ' + sender);
        // console.log('Last Message: ' + lastMessage);
        // console.log('Sender Image URL: ' + senderImageURL);
        // console.log('Date: ' + date);
    };

    return (
        <a onClick={handleClick} href="/room/!MxHlh0MiugoGF9z1iOaB1y8SRI2bpAMCkLPGAQi0B40:reddit.com"
            className="relative box-border flex min-h-[60px] flex-row items-center gap-[var(--spacer-xs)] border-[none] bg-[color:var(--color-tone-7)] px-[var(--spacer-sm)] py-[var(--spacer-xs)] text-left no-underline hover:bg-[var(--color-secondary-background-hover)] active:bg-[var(--color-interactive-pressed)]">
            <span className="box-border size-[var(--size-xl)]">
                <span className="inline-flex items-center justify-center">
                    <span className="fp-avatar-container relative isolate box-border inline-flex size-8 rounded-full">
                        <span style={{background: 'var(--color-avatar-gradient)'}}
                            className="border-neutral-background h-100 w-100 relative box-border justify-center rounded-full border border-solid">
                            <span className="absolute bottom-0 left-0 z-[2] inline-flex size-8 shrink-0 justify-center">
                                <svg width="100%" height="100%" viewBox="0 0 121 122" xmlns="http://www.w3.org/2000/svg" className="overflow-hidden">
                                    <defs>
                                        <clipPath id="circle-clip">
                                            <circle cx="60.5" cy="61" r="60" />
                                        </clipPath>
                                    </defs>
                                    <image height="100%" width="100%" href={senderImageURL} alt="Sender Avatar" clipPath="url(#circle-clip)"></image>
                                </svg>
                            </span>
                        </span>
                    </span>
                </span>
            </span>
            <div className="flex flex-1 flex-col gap-[var(--spacer-4xs)] overflow-hidden">
                <div className="flex min-w-0 justify-between gap-[var(--spacer-xs)]">
                    <span className="truncate text-xs text-[color:var(--color-tone-1)]">{sender}</span>
                    <span className="whitespace-nowrap text-[0.65rem] text-[color:var(--color-tone-2)]">{date}</span>
                </div>
                <div className="flex min-w-0 justify-between gap-[var(--spacer-xs)]">
                    <span className="whitespace-nowrap text-[0.65rem] text-[color:var(--color-tone-2)]">{lastMessage}</span>
                    <div className="flex flex-nowrap items-center gap-[var(--spacer-xs)]"></div>
                </div>
            </div>
            {unreadTotal > 0 && (
                <div className="py-0.5 absolute right-4 top-[55%] -translate-y-1/2 rounded-full bg-blue-500 px-1 text-[8px] leading-none text-white">
                    {unreadTotal}
                </div>
            )}
        </a>
    );
}

ChatPreview.defaultProps = {
    sender: 'Unknown',
    lastMessage: 'No message available',
    senderImageURL: 'https://example.com/default-avatar.png',
    date: 'Unknown Date',
    unreadTotal: 0,
};

ChatPreview.propTypes = {
    sender: PropTypes.string,
    lastMessage: PropTypes.string,
    senderImageURL: PropTypes.string,
    date: PropTypes.string,
    unreadTotal: PropTypes.number,
};

export {ChatPreview};
