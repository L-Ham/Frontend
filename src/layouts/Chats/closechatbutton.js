import React from 'react';
import PropTypes from 'prop-types';
// disable eslint maxline length
/* eslint-disable max-len */

/**
 * The chat preview component.
 * @component
 * @example
 * // Render the chat preview.
 * <LeftBar />
 * @return {JSX.Element} The chat preview component.
 */
function CloseChatButton({minimizeChat}) {
    return (

        <button onClick={minimizeChat} aria-label="Minimize chat" className="
        button-medium button-plain
        icon button inline-flex items-center
        justify-center rounded-full p-1
        text-[color:var(--color-tone-2)] hover:bg-[var(--color-secondary-background-hover)]
        active:bg-[var(--color-interactive-pressed)]" type="button">
            <span className="flex items-center justify-center">
                <span className="flex">
                    <svg fill="currentColor"
                        height="16"
                        viewBox="0 0 20 20" width="16"
                        xmlns="http://www.w3.org/2000/svg">
                        <path d="M10 13.125a.624.624 0 0 1-.442-.183l-5-5 .884-.884L10 11.616l4.558-4.558.884.884-5 5a.624.624 0 0 1-.442.183Z"></path>
                    </svg></span>   </span>      </button>
    );
}
// write prop validation
CloseChatButton.propTypes = {
    minimizeChat: PropTypes.func,
};
export {CloseChatButton};
