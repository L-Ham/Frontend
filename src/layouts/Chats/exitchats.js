import React from 'react';

// disable eslint maxline length
/* eslint-disable  */

/**
 * The chat preview component.
 * @component
 * @example
 * // Render the chat preview.
 * <LeftBar />
 * @return {JSX.Element} The chat preview component.
 */
function ExitChats({show}) {
    return (

        <button onClick={()=>{
            show(false);
        }} id = 'exitchats' aria-label="Minimize chat" className="
        button-medium button-plain
        icon button inline-flex items-center
        justify-center rounded-full p-1
        text-[color:var(--color-tone-2)] hover:bg-[var(--color-secondary-background-hover)]
        active:bg-[var(--color-interactive-pressed)]" type="button">
            <span className="flex items-center justify-center">
                <span className="flex">
                    <svg fill="currentColor" height="16" viewBox="0 0 20 20" width="16" xmlns="http://www.w3.org/2000/svg">
                        <path d="m18.442 2.442-.884-.884L10 9.116 2.442 1.558l-.884.884L9.116 10l-7.558 7.558.884.884L10 10.884l7.558 7.558.884-.884L10.884 10l7.558-7.558Z"></path> </svg></span>   </span>      </button>
    );
}

export {ExitChats};
