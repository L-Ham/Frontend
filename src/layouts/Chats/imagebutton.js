import React, {useRef} from 'react';
/*eslint-disable*/
function ImageButton({ setImage, setPreview }) {
    const fileInputRef = useRef(null);

    const handleButtonClick = () => {
        fileInputRef.current.click();
    };

    const handleFileChange = (event) => {
        const file = event.target.files[0];
        if (file) {
            const imageUrl = URL.createObjectURL(file);
            setPreview(imageUrl); // Set image URL for preview
            setImage(file); // Set image file for sending
        }
        event.target.value = null; // Clear the input to allow the same file to be re-selected if needed
    };

    return (
        <>
            <button
                className="button-medium button-plain icon button inline-flex items-center justify-center rounded-full p-2 text-[color:var(--color-tone-2)] hover:bg-[var(--color-secondary-background-hover)] active:bg-[var(--color-interactive-pressed)]"
                type="button"
                onClick={handleButtonClick}
            >
                <span className="flex items-center justify-center">
                    <span className="flex">
                        <svg fill="currentColor" height="20" viewBox="0 0 20 20" width="20" xmlns="http://www.w3.org/2000/svg">
                            <path d="M10.007 14a4 4 0 1 1 0-8 4 4 0 0 1 0 8Zm0-6.75a2.75 2.75 0 1 0 0 5.5 2.75 2.75 0 0 0 0-5.5ZM17.375 18H2.625A2.63 2.63 0 0 1 0 15.375V5.656a2.629 2.629 0 0 1 2.625-2.625h2.389L6.23 1.454A1.166 1.166 0 0 1 7.153 1h5.739a1.155 1.155 0 0 1 .907.446l1.222 1.585h2.355A2.627 2.627 0 0 1 20 5.656v9.718A2.629 2.629 0 0 1 17.375 18ZM2.625 4.281A1.377 1.377 0 0 0 1.25 5.656v9.719a1.377 1.377 0 0 0 1.375 1.375h14.75a1.377 1.377 0 0 0 1.375-1.375V5.657a1.375 1.375 0 0 0-1.374-1.375h-2.969l-1.6-2.073-5.654.041-1.524 2.031H2.625Z"></path>
                        </svg>
                    </span>
                </span>
            </button>

            {/* Hidden file input */}
            <input
                ref={fileInputRef}
                type="file"
                accept="image/*"
                style={{ display: 'none' }}
                onChange={handleFileChange}
            />
        </>
    );
}

export {ImageButton};
