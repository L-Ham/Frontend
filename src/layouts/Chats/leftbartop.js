import React, {useState} from 'react';
import PropTypes from 'prop-types';
/*eslint-disable*/
function LeftBarTop({ addingUser,filterChats }) {
    const [menuVisible, setMenuVisible] = useState(false);
    const [dropdownVisible, setDropdownVisible] = useState(false);

    const toggleMenu = () => setMenuVisible(!menuVisible);
    const toggleDropdown = () => setDropdownVisible(!dropdownVisible);

    const handleMenuClick = (option) => {
        console.log(option);
        filterChats(option) // Handle the option here or propagate it up
        setDropdownVisible(false); // Hide the dropdown after selection
    };

    return (
        <div className="flex items-center px-4 py-3">
            <span className="text-lg font-bold leading-6">Chats</span>
            <button onClick={() => {
                toggleMenu();
                addingUser();
            }}
            className="ml-auto mr-1 box-border
                inline-flex h-[var(--button-height)] cursor-pointer items-center justify-center truncate rounded-[999px] border-[solid_var(--button-border-color,transparent)] px-[var(--rem6)] text-center leading-[calc(var(--button-height)_-_var(--button-border-width,0px)*2)] text-[color:var(--button-color-text)] no-underline shadow-[var(--button-shadow)] hover:bg-[var(--color-secondary-background-hover)] active:bg-[var(--color-interactive-pressed)]"
            >
                <span className="flex items-center justify-center">
                    <svg fill="currentColor" height="20" viewBox="0 0 20 20" width="20" xmlns="http://www.w3.org/2000/svg">
                        <path d="M17.063 2.931A9.99 9.99 0 0 0 .123 8.444a9.883 9.883 0 0 0 1.195 6.49L.085 19.009a.729.729 0 0 0 .9.913l4.166-1.194a9.856 9.856 0 0 0 6.448 1.142 9.989 9.989 0 0 0 8.12-12.214 9.991 9.991 0 0 0-2.656-4.725Zm1.57 8.499a8.784 8.784 0 0 1-7.227 7.2 8.664 8.664 0 0 1-5.856-1.112l-.231-.139-3.762 1.078 1.118-3.691-.145-.238a8.655 8.655 0 0 1-1.172-5.893 8.751 8.751 0 1 1 17.275 2.8v-.005Zm-8.008-2.058h3.374v1.25h-3.375v3.379h-1.25v-3.376H6v-1.25h3.375V6.002h1.25v3.37Z" />
                    </svg>
                </span>
            </button>

            <div className="relative">
                <button
                    id="1"
                    className="ml-1 button-medium button-plain
                    icon button inline-flex items-center
                    justify-center rounded-full p-1
                    text-[color:var(--color-tone-2)] hover:bg-[var(--color-secondary-background-hover)]
                    active:bg-[var(--color-interactive-pressed)]"
                    aria-haspopup="true"
                    aria-expanded={dropdownVisible ? 'true' : 'false'}
                    onClick={toggleDropdown}
                >
                    <span className="flex items-center justify-center">
                        <svg fill="currentColor" height="20" viewBox="0 0 20 20" width="20" xmlns="http://www.w3.org/2000/svg">
                            <path d="M10,13.125a.624.624 0 0 1-.442-.183l-5-5 .884-.884L10,11.616l4.558-4.558.884.884-5,5a.624.624 0 0 1-.442.183Z" />
                        </svg>
                    </span>
                </button>

                {dropdownVisible && (
                    <div className="absolute right-0 top-full mt-1 py-1 w-48 bg-white rounded-md shadow-lg z-50">
                        <button onClick={() => handleMenuClick('all')} className="block px-4 py-2 text-sm text-gray-700 hover:bg-gray-100 w-full text-left">All</button>
                        <button onClick={() => handleMenuClick('single')} className="block px-4 py-2 text-sm text-gray-700 hover:bg-gray-100 w-full text-left">Single</button>
                        <button onClick={() => handleMenuClick('group')} className="block px-4 py-2 text-sm text-gray-700 hover:bg-gray-100 w-full text-left">Group</button>
                    </div>
                )}
            </div>
        </div>
    );
}

LeftBarTop.propTypes = {
    addingUser: PropTypes.func.isRequired,
};

export { LeftBarTop };
