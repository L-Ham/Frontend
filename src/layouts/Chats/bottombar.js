import React, {useState} from 'react';
import {ImageButton} from './imagebutton';
import {MessageBox} from './messagebox';
import {SendMessageButton} from './sendmessagebutton';
/*eslint-disable*/
function BottomBar({sendText, sendImage}) {
    const [textMessage, setTextMessage] = useState('');
    const [imageMessage, setImageMessage] = useState(null); // for sending
    const [imagePreviewUrl, setImagePreviewUrl] = useState(null); // for display

    const handleSendMessage = async (type = 1) => {
        if (type === 1 && textMessage.trim() !== '') {
            sendText(textMessage);
            setTextMessage(''); // Reset text message
        } else if (type === 2 && imageMessage) {
            sendImage(imageMessage);
            setImageMessage(null); // Reset image message
            setImagePreviewUrl(null); // Also reset the image preview URL
        }
    };

    const handleRemoveImage = () => {
        setImageMessage(null);
        setImagePreviewUrl(null);
    };

    return (
        <div>
            {imagePreviewUrl && (
                <div className="relative flex items-center justify-center border border-gray-300 bg-white p-2">
                    <img id = 'img1' src={imagePreviewUrl} alt="Preview" className="max-h-[100px]" />
                    <button id = 'ab' onClick={handleRemoveImage} className="absolute right-0 top-0 rounded-full bg-white p-1">
                        &#x2715;
                    </button>
                </div>
            )}
            <form id = '3'
                className='flex gap-[var(--spacer-2xs)] p-[var(--spacer-xs)]'
                onSubmit={(e) => e.preventDefault()}
            >
                <ImageButton setImage={setImageMessage} setPreview={setImagePreviewUrl} />
                <MessageBox setMessage={setTextMessage} message={textMessage} />
                <SendMessageButton sendMessage={handleSendMessage} image={imageMessage} />
            </form>
        </div>
    );
}

export {BottomBar};
