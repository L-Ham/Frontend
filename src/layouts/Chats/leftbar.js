import React from 'react';
import {LeftBottom} from './leftbottom';
import {LeftThreads} from './leftthreads';
import {LeftBarTop} from './leftbartop';
import PropTypes from 'prop-types';

// disable eslint maxline length
/* eslint-disable max-len */

/**
 * The chat preview component.
 * @component
 * @example
 * // Render the chat preview.
 * <LeftBar />
 * @return {JSX.Element} The chat preview component.
 */
function LeftBar({chatsToLeft, onSelectChat, addUser, filter}) {
    return (
        <div className='box-border flex h-full flex-col overflow-auto border-r border-solid border-r-[color:var(--color-tone-5)]'>
            <LeftBarTop addingUser = {addUser} filterChats = {filter} />
            <LeftThreads onSelect = {onSelectChat} chats = {chatsToLeft}/>
            <LeftBottom chatsToBottom = {chatsToLeft} onSelectChat={onSelectChat}/>

        </div>

    );
}
// write prop validation
LeftBar.propTypes = {
    chatsToLeft: PropTypes.array,
    onSelectChat: PropTypes.func,
    addUser: PropTypes.func,
    filter: PropTypes.func,
};
export {LeftBar};
