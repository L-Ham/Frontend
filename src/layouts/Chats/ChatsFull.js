import React, {useState, useEffect, useRef} from 'react';
import {BottomBar} from './bottombar';
import {MessageContainer} from './messages/messagecontainer';
import {TopBar} from './topbar';
import {LeftBar} from './leftbar';
import {Message} from './messages/message';
import {AddChat} from './new chat/addchat';
import {axiosInstance} from '../../requests/axios';
import {API_ROUTES} from '../../requests/routes';
import {io} from 'socket.io-client';

/* eslint-disable*/
 
 import {useSelector} from 'react-redux';

 //import {useSelector} from 'react-redux';
//const token = useSelector((state) => state.use.tokenr);
 





function ChatsFull({show}) {
    // Define default avatar and the conversion functions outside the component or hook logic to avoid re-creation on each render
    const defaultAvatar = 'https://i.redd.it/lif0817mz3951.png';
    const [message, setMessage] = useState('');
    const ourtoken = useSelector((state) => state.user.token);
    const socket = useRef(null);
    const [myAvatar, setMyAvatar] = useState(defaultAvatar);
    const [selectedChatId, setSelectedChatId] = useState('');
    const selectedChatIdRef = useRef(selectedChatId);
    const [filteredChats, setFilteredChats] = useState([{}]);
    const [filterType,setFilterType] = useState('all');
    const [receivedMessages, setReceivedMessages] = useState([]);
    const username = useSelector((state) => state.user.username);
    const [isFullScreen, setIsFullScreen] = useState(false);
    console.log('Username:', username);
    const addNewMessageToChat = (chatId, newMessage) => {
        setChats(prevChats => {
            const updatedChats = {...prevChats};
            // Check if the chat exists
            if (updatedChats[chatId]) {
                // Push the new message into the chat's messages array
                updatedChats[chatId].messages.push(newMessage);
                console.log('Adding the message', updatedChats[chatId].messages);
    
                // If the chat is of type 'group', update the chat's avatar to the newMessage's avatar
                if (updatedChats[chatId].type === 'group') {
                    updatedChats[chatId].avatar = newMessage.avatar;
                    console.log('Updated group chat avatar to', newMessage.avatar);
                }
            } else {
                console.error("Chat ID not found, unable to add message:", chatId);
                // Optionally, you could initialize a new chat here if that's intended behavior
            }
            return updatedChats;
        });
    };
    

    async function fetchAvatar() {
        try {
            const response = await axiosInstance.get(API_ROUTES.avatar);
            // Directly use response.data since it matches the expected structure
            setMyAvatar(response.data.url);
        } catch (error) {
            if (error?.response?.data?.message === 'Avatar image not found') return;
            console.error('Failed to fetch banner:', error);
        }
    }
    useEffect(() => {
        fetchAvatar();
    }, []);
    //make filtered chats have the chats that are filtered





    
    useEffect(() => {
        selectedChatIdRef.current = selectedChatId;
    }, [selectedChatId]);
    
    //const [socket, setSocket] = useState([]);
    useEffect(() => {
        if (!socket.current) {
            console.log('Connecting to server with token:', ourtoken);
            socket.current = io('http://api.reddit-bylham.me', {
                query: { token: `Bearer ${ourtoken}` },
            });
            //past: 
    
            const handleNewMessage = (message) => {
                console.log('Received new message:', message);
                const formattedMessage = {
                    time: convertDateFormat(message.createdAt),
                    sender: message.senderName,
                    message: message.message || "",
                    imgsrc: message.imageUrl,
                    avatar: message.senderAvatar,
                    type: message.type,
                    isRead: message.isRead,
                    id: message._id
                };
                console.log('Formatted message based on type:', formattedMessage);
                console.log('Adding message to chat:', formattedMessage);


                addNewMessageToChat(message.conversationId, formattedMessage);
            };
    
            socket.current.on('newMessage', handleNewMessage);
        }
    
        return () => {
            if (socket.current) {
                socket.current.off('newMessage');
                socket.current.disconnect();
                socket.current = null;
            }
        };
    }, [ourtoken]);  // Only re-run if `ourtoken` changes
     // Dependency on ourtoken to reconnect if it changes
    
    
    const toggleNow = () => {
        setNow(!now);
    }

    
    // Event listener for successfully connecting to the server

  
    function convertDateFormat(dateString) {
        const now = new Date();
        const date = new Date(dateString);
        const diffInSeconds = Math.round((now - date) / 1000);
        const diffInMinutes = Math.round(diffInSeconds / 60);
        const diffInHours = Math.round(diffInMinutes / 60);
        const diffInDays = Math.round(diffInHours / 24);

        if (diffInDays > 0) {
            return `${diffInDays} days ago`;
        } else if (diffInHours > 0) {
            return `${diffInHours} hours ago`;
        } else if (diffInMinutes > 0) {
            return `${diffInMinutes} minutes ago`;
        } else {
            return `now`;
        }
    }
    function convertData(data, username) {
        const chats = {};
        data.conversations.forEach((conversation) => {
            const chatId = conversation._id;
            let chatName = conversation.chatName; // Default to using the predefined chatName
            let Type = conversation.type;
    
            // Attempt to set the primaryAvatar to the avatar of the last message sender
            const lastMessage = conversation.messages.slice(-1)[0];
            let primaryAvatar = lastMessage ? conversation.participantsAvatarUrls[conversation.participants.indexOf(lastMessage.senderName)] : conversation.participantsAvatarUrls[0];
            console.log('primary avatar:', primaryAvatar);
    
            // Check if the conversation type is 'single' and has exactly two participants
            if (conversation.type === 'single' && conversation.participants.length === 2) {
                const otherParticipantIndex = conversation.participants.findIndex(participant => participant !== username);
                chatName = conversation.participants[otherParticipantIndex]; // Find the other participant's name
                primaryAvatar = conversation.participantsAvatarUrls[otherParticipantIndex]; // Use the other participant's avatar
                console.log('Chat name:', chatName);
                console.log('Username:', username);
            }
    
            const avatarUrls = conversation.participantsAvatarUrls.map((avatar) => avatar || defaultAvatar);
    
            const messages = conversation.messages.map((message) => ({
                time: convertDateFormat(message.createdAt),
                sender: message.senderName,
                message: message.message || 'Image',
                imgsrc: message.imageUrl || null,
                type: message.type,
                avatar: avatarUrls[conversation.participants.indexOf(message.senderName)] || defaultAvatar,
            }));
    
            chats[chatId] = {
                name: chatName, // Use the dynamically determined chatName
                avatar: primaryAvatar, // Use the avatar determined above
                messages: messages,
                unreadCount: conversation.unreadCount,
                type : Type,
            };
        });
        console.log('Chats:', chats);
        return chats;
    }
    
    
    
    
   /* useEffect(() => {
        // Listen for messages
        socket.on('newMessage', (message) => {
            console.log('Received message:', message);
        });

        // Cleanup on component unmount
        return () => {
            socket.off('newMessage');
        };
    }, []);*/
    const sendTextMessage = async (textMessage) => {
        const messageData = {
            chatId: selectedChatId,
            type: "text",
            message: textMessage,
        };
        console.log('Sending message:', messageData);
    
        try {
            const response = await axiosInstance.post(API_ROUTES.sendChatMessage, messageData);
            console.log('Message sent successfully:', response.data);
            addNewMessageToChat(selectedChatId, {
                sender: username, // Adjust according to your app's context
            type: "text",
            message: textMessage,
            avatar: myAvatar,

            });
            console.log('image: ',myAvatar);
        } catch (error) {
            console.error('Failed to send message:', error);
        }
    };
    const sendImageMessage = async (image) => {
        const formData = new FormData();
        formData.append('file', image);
        formData.append('chatId', selectedChatId);
        formData.append('type', 'image');
    
        console.log('Sending image:', formData);
    
        try {
            const response = await axiosInstance.post(API_ROUTES.sendChatMessage, formData, {
                headers: {'Content-Type': 'multipart/form-data'}
            });
            console.log('Image sent successfully:', response.data);
    
            // Add new message to chat history in state
            addNewMessageToChat(selectedChatId, {
                sender: username, // Adjust according to your app's context for sender identification
                type: "image",
                message: "", // Since it's an image, the message could be empty or an image description if you have it
                imgsrc: URL.createObjectURL(image), // This assumes you want to display the image immediately before any server 
                avatar: myAvatar,
            });
        } catch (error) {
            console.error('Failed to send image:', error);
        }
    };
    
    
     // const sendMessage = () => {
       // socket.emit('sendMessage', message);
       // setMessage('');
       // console.log('Sent message:', message);
      //};

    const [chats, setChats] = useState({});
    const [selectedChat, setSelectedChat] = useState({messages: []});
    const [now, setNow] = useState(false);

    const [addUser, setAddUser] = useState(false);
    const [chatsHidden, setChatsHidden] = useState(false);
   
    const toggleAddUser = () => {
        setAddUser(!addUser);
    };
    function minimize() {
        setChatsHidden(true);
    }
    function expand() {
        setIsFullScreen(!isFullScreen);
    }
    
    useEffect(() => {
        console.log('chats,,', chats);
        if (filterType === 'all') {
            setFilteredChats(chats);
            console.log('chatssss',chats);
            console.log('filtered chats: ',chats);
        } else {
            const filtered = {};
            Object.keys(chats).forEach((chatId) => {
                if (chats[chatId].type.toLowerCase().includes(filterType.toLowerCase())) {
                    filtered[chatId] = chats[chatId];
                }
            });
            setFilteredChats(filtered);
            console.log('filtered: ', filtered);
        }
    }, [chats, filterType]);
    

    
        
    useEffect(() => {
        async function fetchAllChats() {
            try {
                const response = await axiosInstance.get(API_ROUTES.getAllChats);
                console.log('Chats response:', response);
                const newChats = convertData(response.data,username);
                console.log('newChats: ',newChats);
                setChats(newChats);
                if (Object.keys(newChats).length > 0) {
                    setSelectedChat(newChats[Object.keys(newChats)[0]]);
                    setSelectedChatId(Object.keys(newChats)[0]);
                    console.log('Selected chat id:', Object.keys(newChats)[0]);
                    console.log('Selected chat:', newChats[Object.keys(newChats)[0]]);
                }
            } catch (error) {
                console.error('Failed to fetch chats:', error);
            }
        }

        fetchAllChats();
    }, [now]);

    const handleChatSelection = (chatName) => {
        setSelectedChat(chats[chatName]);
        setSelectedChatId(chatName);
        console.log('Selected chat id:', chatName);
        console.log(chatName);
    };
    useEffect(() => {
        if (isFullScreen && !chatsHidden) {
            document.body.style.overflow = 'hidden';
        } else {
            document.body.style.overflow = 'auto';  // or 'initial' if that was your default
        }

        // Cleanup function to reset overflow when the component unmounts
        return () => {
            document.body.style.overflow = 'auto';  // or 'initial'
        };
    }, [isFullScreen,chatsHidden]);

    const messageComponents = selectedChat && selectedChat.messages 
    ? selectedChat.messages.map((message, index) => (
        <Message key={index} time={message.time} name={message.sender} message={message.message} avatar={message.avatar} imgsrc={message.imgsrc} type={message.type} />
      )) 
    : [];

    return (
        <div className={isFullScreen ? 'fixed top-0 right-0 bottom-0 left-0' : 'fixed bottom-0 right-8'} style={{ zIndex: 1000 }}>
            <div className='block min-w-0'>
                <div className='overflow-hidden rounded-t-[16px] border border-b-0 border-solid border-[color:var(--color-neutral-border-weak)] bg-[color:var(--color-neutral-background)] shadow-md'>
                    <div className={chatsHidden ? 'hidden' : ''}>
                        <div className={isFullScreen  ? 'h-screen w-full' : 'h-[506px] w-[632px]'}>
                            <div className='grid h-full grid-cols-[220px_1fr] overflow-hidden'>
                                <LeftBar filter = {setFilterType} chatsToLeft={filteredChats} onSelectChat={handleChatSelection} addUser={toggleAddUser} />
                                <div className='h-full overflow-hidden'>
                                    <div className='relative flex h-full overflow-hidden'>
                                        <div className='flex h-full grow flex-col overflow-auto'>
                                            <TopBar name={selectedChat ? selectedChat.name : ''} minimize={minimize} expand = {expand} showing = {show} full = {isFullScreen}  />
                                            {!addUser && <MessageContainer messages={messageComponents} />}
                                            {addUser && <AddChat reverse={() => setAddUser(!addUser)} fetch = {toggleNow} />}
                                            {!addUser && <BottomBar sendText={sendTextMessage} sendImage={sendImageMessage} setChat={setSelectedChat} />}
                                        </div>
                                    </div>
                                </div>
                            </div>
                        </div>
                    </div>
                    {chatsHidden && <div className="group relative">
                        <button onClick={() => setChatsHidden(false)} className="flex min-w-[128px] cursor-pointer items-center border-none bg-transparent py-3 pl-4 pr-12" aria-label="Maximize chat view">
                            <p className="m-0 text-sm leading-4">Chat</p>
                        </button>
                        <button onClick = {()=>{show(false)}} className="button-x-small button-plain icon button absolute right-4 top-2 inline-flex items-center justify-center px-[var(--rem6)] opacity-0 focus:opacity-100 group-hover:opacity-100" data-testid="chat-close-button" type="button">
                            <span className="flex items-center justify-center">
                                <span className="flex">
                                    <svg fill="currentColor" height="12" viewBox="0 0 20 20" width="12" xmlns="http://www.w3.org/2000/svg">
                                        <path d="m18.442 2.442-.884-.884L10 9.116 2.442 1.558l-.884.884L9.116 10l-7.558 7.558.884.884L10 10.884l7.558 7.558.884-.884L10.884 10l7.558-7.558Z"></path>
                                    </svg>
                                </span>
                            </span>
                        </button>
                    </div>}
                </div>
            </div>
        </div>
    );
    
}

export {ChatsFull};