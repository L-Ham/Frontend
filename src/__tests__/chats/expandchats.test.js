import React from 'react';
import {render} from '@testing-library/react';
import {ExpandChatsButton} from '../../layouts/Chats/expandchats';
import {describe, it, jest, expect} from '@jest/globals';
import '@testing-library/jest-dom';
describe('ExpandChatsButton', () => {
    // Test to ensure it renders without crashing
    it('renders without crashing', () => {
        const mockExpandChats = jest.fn();
        const {getByRole} = render(<ExpandChatsButton expandChats={mockExpandChats} />);
        expect(getByRole('button')).toBeInTheDocument();
    });

    // Accessibility test to ensure the button has the correct aria-label
    it('has proper aria-label for accessibility', () => {
        const mockExpandChats = jest.fn();
        const {getByLabelText} = render(<ExpandChatsButton expandChats={mockExpandChats} />);
        expect(getByLabelText('Expand chats')).toBeInTheDocument();
    });

    // Snapshot test to ensure the rendered output does not change unexpectedly
    it('matches snapshot', () => {
        const mockExpandChats = jest.fn();
        const {asFragment} = render(<ExpandChatsButton expandChats={mockExpandChats} />);
        expect(asFragment()).toMatchSnapshot();
    });
});
