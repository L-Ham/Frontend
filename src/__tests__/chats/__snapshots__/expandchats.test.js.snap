// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpandChatsButton matches snapshot 1`] = `
<DocumentFragment>
  <button
    aria-label="Expand chats"
    class="
 button-medium button-plain
 icon button inline-flex items-center
 justify-center rounded-full p-1
 text-[color:var(--color-tone-2)] hover:bg-[var(--color-secondary-background-hover)]
 active:bg-[var(--color-interactive-pressed)]"
    id="expand"
    type="button"
  >
    <span
      class="flex items-center justify-center"
    >
      <span
        class="flex"
      >
        <svg
          fill="currentColor"
          height="16"
          viewBox="0 0 20 20"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15.75 13H17v3.375A2.63 2.63 0 0 1 14.375 19H3.625A2.63 2.63 0 0 1 1 16.375V5.625A2.629 2.629 0 0 1 3.625 3H7v1.25H3.625A1.377 1.377 0 0 0 2.25 5.625v10.75a1.377 1.377 0 0 0 1.375 1.375h10.75a1.377 1.377 0 0 0 1.375-1.375V13Zm2.625-12h-5v1.25h3.491l-8.433 8.433.884.884 8.433-8.433v3.491H19v-5A.625.625 0 0 0 18.375 1Z"
          />
        </svg>
      </span>
    </span>
  </button>
</DocumentFragment>
`;
